Extension types page

MaryaBelanger opened the pull request:
Fixes #4177 

---

The [first commit](https://github.com/dart-lang/site-www/pull/5508/commits/2264edc5b7f00e2d2080e423b812472949af9d4a#diff-b9da11e0de118f5b3934ae1dd3ad2c67a33fa1ad98188a6603d53da1435e6004R270) includes a "Use cases" section. I removed this from the second commit as I felt like the information covered there is more generally covered across the rest of the page where syntax and usage are discussed. 

But, if anyone thinks the use cases are valuable I am happy to include them (I didn't fully flesh out those sections because I was leaning towards removing them already, but if we want to keep them I will improve/refine them)

**Link to page:** https://dart-dev--pr5508-extension-types-6op3ayna.web.app/language/extension-types

---

TODO:

- After some offline feedback from @domesticmouse, justify the functionality so readers aren't left asking "why?"
  - make it clearer up front that this feature's main purpose is enabling JS interop, 
  - maybe use interop code as the top sample / maybe even have all the samples in a JS context?

- Maybe take the ["Use cases"](https://github.com/dart-lang/site-www/pull/5508/commits/2264edc5b7f00e2d2080e423b812472949af9d4a#diff-b9da11e0de118f5b3934ae1dd3ad2c67a33fa1ad98188a6603d53da1435e6004R270) from the first commit and write a blog post? 
  - Erik also gave me an interesting scenario about wrapping objects from a typeless tree data structure and the performance benefits extension types would have, which could be another interesting scenario to pad out the blog post

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,314 @@
+like the one `implements` introduces for classes,
+but rather an [applicability][] relationship like that between an
+extension method and its `on` type.
+Members that apply to the supertype are applicable to the subtype,
+unless the supertype has a more specific definition for the member.
```
@eernstg I'm not sure if this statement is correct (or worth mentioning?)

lrhn commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+```
+
+Then you can invoke members on the object as you would with a class object.
+Any attempt to treat the extension type object as an object
````
"Any attempt" is a little ambitious.
I'd usually include `o as RepresentationType` in ways to treat an object as `RepresentationType, and that'll work fine.
Is this trying to say: *Can't assign extension type to representation type, or invoke representation type members on extension type, unless the extension type implements the representation type. Can't assign representation type to extension type ever (unless representation type is `dynamic`).

If that's the point, then I think it has been mostly made above already, but summarizing can be useful.

Maybe:
> Subtyping and assignability works normally for extension
> types, which are (statically) treated as completely new types,
> subtypes of `Object?` and supertypes of `Never`, 
> and otherwise only related to other types through their
> `implements` clauses.
> An extension type cannot be assigned to its representation
> type, and it doesn't expose the representation type's 
> members, not unless the extension type explicitly implements
> the representation type to make itself be a subtype of the 
> representation type (see [transparency](#transparency)). 
>
> Likewise, the representation type of an extension type
> is never a subtype of the extension type, so it cannot
> be used where the extension type is needed.

MaryaBelanger replied:
> If that's the point, then I think it has been mostly made above already,

That is the point, and I think you're right. There's a lot of ideas repeated in different sections of this doc I should go through and minimize.

lrhn commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+Any attempt to treat the extension type object as an object
+of the representation type, or vice versa, will result in compile time errors
+(unless that behavior is explicitly defined for the extension type,
+or it is [transparent](#transparency)).
```
Not sure how it can be explicitly defined *other* than being "transparent", so this paragraph confuses me.

Consider (if not taking the suggestion above):
> (unless the extension type implements the representation type, then it can be used as having the representation type).

lrhn commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+
+```dart
+void testE() { 
+  var num = NumberE(1);
````
Call this `num1`. Don't want to get confused with the type `num`.

mnordine replied:
Was thinking similar (but `n`). `num` is used as a variable name several times in the code samples.

MaryaBelanger replied:
Oof, big oversight on my end. Thanks for catching!

lrhn commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
+    1 as NumberE, // Ok: Dynamic type cast recognizes representation type.
```
Not sure what "recognizes" means.

It's just OK, because it has static type `NumberE`.
It would also be (statically) OK to write `"nope" as NumberE`. That would fail at runtime, but in an example with multiple compile-time errors, runtime is a foreign country. I wouldn't expect the program to be about runtime behavior at all.

(Maybe split the example into two: One with errors, and one with correct usages. Then the reader will expect the correct usages to be *runtime* correct too.)

MaryaBelanger replied:
Ok, it sounds like the `as` cast example just doesn't fit here so I'll remove it

lrhn commented on src/language/keywords.md:
```diff
@@ -16,23 +16,23 @@ The following table lists the words that the Dart language treats specially.
+| [assert][]            | [export][]{{bii}}         | [in][]                | [static][]{{bii}}  |
+| [async][]{{ckw}}      | [extends][]               | [interface][]{{bii}}  | [super][]          |
+| [await][]{{lrw}}      | [extension][]{{bii}}      | [is][]                | [switch][]         |
+| [base][]{{bii}}       | [extension type][]{{bii}} | [late][]{{bii}}       | [sync][]{{ckw}}    |
```
(This marks both `extension` and `extension type` as built-in identifiers. Since `extension type` is not an identifier - it's two - it can't be a built-in identifier. I think `type` might be a contextual keyword, because it's only a keyword after `extension` when `extension` occurs as a declaration.)

MaryaBelanger replied:
Hmm, not sure how to handle this. I could change it to just:

> `| [type][]{{ckw}} |`

Or

> `| [(extension) type][]{{ckw}} |`

The same way we have parens acround (class) and (variable) for `final`, but I'm not sure if that fixes it. @parlough

eernstg replied:
`extension` is a built-in identifier, but `type` is not. You could say that `type` is a contextual built-in identifier (or a contextual reserved word, because the 'reserved word'/'built-in identifier' distinction makes no difference here, and we might want to describe `'type'` in the same way that we currently describe `'await'` and `'yield'`).

The required context is that we're parsing a `<topLevelDefinition>` and we have accepted the token `'extension'`. In that situation the token `'type'` is a reserved word that implies we're parsing an `<extensionTypeDeclaration>`, not an `<extensionDeclaration>`.

I think `'type'` could then have its own entry of the form `[type][]{{lrw}}`.

MaryaBelanger replied:
Thank you! That's a tricky one but I'm glad we could get it right

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  NumberE operator +(NumberE other) =>
+      NumberE(value + other.value);
+  // Getter:
+  NumberE get i => value;
```
Perhaps?:
```suggestion
  NumberE get i => this;
```

We could also return `NumberE(value)`, but `this` is already available, and it is the same object as `value`, and it has the required static type, so why not use it?

MaryaBelanger replied:
Sgtm, thanks!

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  var num = NumberE(1);
+  int num2 = NumberE(2); // Error: Can't assign 'NumberE' to 'int'.
+  
+  num.isValid(); // Ok: Extension member invocation.
```
```suggestion
  num.isValid(); // Ok: Extension type member invocation.
```

I think we have to reserve the phrase 'extension member' to the ones that occur in an `extension` declaration, not an `extension type` declaration.

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  List<NumberE> numbers = [
+    NumberE(1), 
+    num.i, // Ok: 'i' getter returns type 'NumberE'.
+    1, // Error: Can't assign 'int' element to list type 'NumberE'.
```
Perhaps?:
```suggestion
    1, // Error: Can't assign an 'int' to the list element type 'NumberE'.
```

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+
+An extension type that [implements](#implements) its representation type
+is considered *transparent* because it gives the extension type access
+to *all* members of the representation type. Otherwise,
```
```suggestion
to *all* members of the representation type that aren't redeclared. Otherwise,
```

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+void main () {
+  var num = NumberT(1);
+  
+  // All OK: Representation type interface is available to extension type:
```
Perhaps?:
```suggestion
  // All OK: `implements int` allows using `int` members on the extension type:
```

The point is that the decision is made based on the type in the implements clause. If we had written `implements num` then we would not be able to invoke `isEven` on an expression of type `NumberT`, even in the case where the representation type is `int`.

There is an extra constraint (that you already mention elsewhere): If we have `implements T` where `T` is a non-extension type then it must be true that the representation type `R` is a subtype of `T`.

It's probably going to be rather common that an extension type uses the same non-extension type `T` as (1) the representation type and (2) as a superinterface (as in `implements ...T...`), but there are other possibilities.

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+
+Transparency is a conceptual decision for the extension type creator to consider.
+A transparent extension type provides an *extended* interface to an existing type.
+You can invoke all the members of the representation type,
```
```suggestion
You can invoke all the members of the representation type that aren't redeclared,
```

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+A transparent extension type provides an *extended* interface to an existing type.
+You can invoke all the members of the representation type,
+plus any auxillary members you declare. The new interface is available
+to instances of the extension type.
```
```suggestion
to expressions whose static type is the extension type.
```

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+## Type considerations
+
+Extension types are a compile-time wrapping construct.
+So, at runtime, there is absolutely no trace of the extension type.
```
```suggestion
So, at run time, there is absolutely no trace of the extension type.
```

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  }
+  
+  switch (num) {
+    case (int x): print(x);  // Matches because runtime type is representation type.
```
Maybe omit the parens:
```suggestion
    case int x: print(x);  // Matches because runtime type is representation type.
```

It's just going to make some readers speculate about why those parentheses are included, and whether it's even allowed. ;-)

eernstg reviewed (approved):
Looks great! I added a bunch of comments.

mnordine commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
```
```suggestion
A constructor must be named, because the representation declaration
```

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  title: Callable objects
+---
+
+An extension type is a compile time abstraction that "wraps"
```
> Of course, please just ignore this comment if this is a discussion that you've had already a long time ago

Not at all! I wasn't sure while writing the whole page (and it comes up a lot). I probably jumped between all variations so now I can go back and fix that in an informed way. Not a pet peeve in my opinion, very useful to know :)

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+to objects of the underlying type, called the *representation type*.
+You can choose to use some members, omit others,
+replace others, and add new functionality, 
+all while getting compile time errors before runtime errors.
```
Yes that was the intention, but now I think the emphasis on "compile-time" and "static-only" in the first sentence of the same paragraph make this concept obvious enough already, so I'm removing this clause

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+an existing type with a different, static-only interface.
+They allow you to enforce discipline on the set of operations available
+to objects of the underlying type, called the *representation type*.
+You can choose to use some members, omit others,
```
Changed to:

> When defining the interface of an extension type,
you can choose to reuse some members of the representation type,
omit others, replace others, and add new functionality.

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+It's in scope in the extension type body, and
+you can access it using its name as a getter:
+
+- Within the extension type body: `this.i`.
```
> In fact, I think our lints will likely suggest removing the this.

Just checked, that is true. Thanks

MaryaBelanger replied:
@lrhn would it make sense to say:

> - Within the extension type body: `i` or `this.i` in a constructor.

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+### Constructors
+
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
```
Is it something to do with the `Name._` syntax that allows you to name the primary constructor, or is that just an arbitrary detail you and @lrhn happened to include in your named primary constructor examples? (I don't know what the `._` does there)

eernstg replied:
Every constructor has a name. In a class/enum/extension-type/... declaration named `D`, a constructor can have the name `D`. Other constructors must then have a different name (in order to avoid name clashes), and we can use anything of the form `'D' '.' <identifier>` to do that, for example, `D.myName`.

With that in mind, some people like to say that the constructor named `D` is "unnamed", "nameless", "anonymous", or something like that, and they say that constructors with a name of the form `D.myName` are "named".

I prefer to just talk about the constructor and its name, never referring to any of them being "named" or "unnamed".

So the primary constructor has the name `E._` when it's declared as in `extension type E._(...) ...`, and it has the name `E` when it is declared as in `extension type E(...) ...`.

The reason why both Lasse and I chose `_` is probably that this is the shortest possible private name, and the whole point of using `extension type E._(...) ...` is that we _don't_ want to provide a constructor which is unconditionally creating an expression of the extension type based on an expression of the representation type.

In other words, we don't want to leave the door completely open for an introduction of the extension type, we want to have more control. For example:

```dart
int _currentIdNumber = 0;

extension type IdNumber._(int _value) {
  IdNumber(): this(_currentIdNumber++);
}
```

The intended and well-behaved approach is to call the public constructor with the name `IdNumber`, and that one does not accept any arguments because it is intended to simply use the next natural number that hasn't been used so far. It is still possible to do `42 as IdNumber`, but we should all know that this is a bad idea, and the fact that we can make the primary constructor private helps communicating this rule.

MaryaBelanger replied:
Thank you very much for the detailed explanation!

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+You can optionally declare constructors in an extension type's body.
+A constructor must be named, because the the representation declaration
+already acts as an unnamed constructor for the extension type
+that initializes the represenation object.
```
@lrhn Sorry, I can't tell if this example is showing something different from the basic example in @eernstg's comment on the same thing. I mean there are obvious differences like the type parameters and making it `const`, but are those details relevant to naming the representation declaration, or did you just happen to include them because they were part of the actual extension type you wrote? Do you think we should include them here?

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+
+```dart
+// 'E' has a named constructor 'n',
+// as well as its built-in unnamed constructor
````
Oh I see. This seems like something to add to the declaration syntax section, that would clarify a lot of things. I will come back to this

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+void main() {
+  var named = E.n(3, "Hello");
+  var unnamed = E(4);
+}
```
Thank you very much!

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
```
Great catch, thanks!

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+void main() {
+  var num = NumberE(1);
+  
+  if (num case int x) {  // True because runtime type is representation type.
```
Thank you, that's much better

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+  switch (num) {
+    case (int x): print(x);  // Matches because runtime type is representation type.
+  }
+}
```
I added a lead-in to this example saying:
> Similarly, the static type of the matched value is that of the extension type
in this example:

But I don't think that's makes sense and I don't really understand the example. What are we showing here? Is it just another way to show that `NumberE` is `int` at run time? It doesn't seem like  "the static type of the matched value (is that `i`?) is `NumberE`" to me, but rather that the runtime type of `NumberE` is `int`

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+}
+```
+
+Members of the representation type are not available to the extension type
````
Used the longer form, thanks

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+You also can define new members unrelated to the representation type,
+like the `i` getter and `isValid` method.
+
+Extension types can't declare instance fields (unless [`external`][]),
```
Changed to "instance variables"!

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+```
+
+An extension type can only implement its representation type
+or a supertype of its representation type.
````
I'm ok with adding that

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+and the reference to the *representation object* is named `i`.
+
+The representation object gives access to the underlying type,
+the same way declaring a final instance variable of the underlying type
```
I changed it to

> The representation object gives the extension type access to an object
typed as the representation type, the same way declaring the type of
a final instance variable as the underlying type would in a wrapper class.

But I'm starting to think removing the second clause might be the best solution because it's a confusing sentence no matter what. I initially included it to clarify the purpose for people who are familiar with wrapper classes, but maybe that's not valuable enough to warrant any possible confusion>

atsansone reviewed (approved):
LGTM! 

Would recommend some language refactoring, but nothing to block this great piece of work!

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+#### `@redeclare`
+
+Replacing a method that shares a name with a member of the supertype
+is not an override relationship for extension types, but a *redeclaration*.
```
I like the explanation but yes probably too much to say here :/ (I still added some of it though)

Where are annotations typically defined? It would be great to have this explanation wherever that it, and point to it from here.

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
````
Oh, ok. I was definitely confusing the lint and the annotation. Thanks!

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+```
+
+If you don't use the annotation, the compiler will choose the [more specific][]
+definition wherever the name is invoked, which might not be the behavior you expect.
````
I had linked to [this section](https://github.com/dart-lang/language/blob/main/accepted/2.7/static-extension-methods/feature-specification.md#specificity) in a previous commit, should I put that back?

I'm starting to think it'd be best to just point to `@redeclare` / `annotate_redeclares` documentation from the "Implements" section, and skip the dedicated `@redelcare` section altogether. The lint documentation isn't very descriptive though. Fixing that could be one option. But where are annotations documented (if at all)? I can't find anything for `@redeclare`

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+}
+```
+
+Transparency is a conceptual decision for the extension type creator to consider.
````
Haha good point!

MaryaBelanger commented on src/language/extension-types.md:
````diff
@@ -0,0 +1,315 @@
+}
+```
+
+#### Transparency
````
If I'm understanding correctly (from this and your other comments on this section) would it be better to say:

"An extension type that implements its representation type, or a supertype of its representation type, can be _thought of_ as transparent, because it provides a kind of overarching access to underlying type members (at least _somewhere_ in the hierarchy / either those of the direct representation type or somewhere higher in the type hierarchy) that is otherwise _not_ the default for the intrinsic isolated nature of extension type interfaces"?

Not that that's exactly what I'll write, but is my understanding correct?

eernstg replied:
> It's also possible to have extension types that implement a _supertype_ of their representation type.

True. I did not push on emphasizing this point because I think it's a natural refinement. The crucial message to communicate is that there are two equally valid, but very different, ways to use an extension type: It can be used (1) to extend the interface of an existing type (adding extra members), and (2) to create an unrelated type with an unrelated interface.

An example of the former is `FancyString` that adds extra methods to `String`; we surely want to be able to use an expression of this type in all the ways that `String` supports. An example of the latter is `IdNumber` that provides an interface which is suitable for ID numbers, but allows us to use plain (and cheap) `int` objects for that purpose; in this case we definitely do not want to silently include the interface of `int` as part of the interface of `IdNumber`.

I think those two modes of operation are so different that they should be established _conceptually_, thus clearly communicating that each of them is OK, but each of them comes with its own conceptualization and software engineering rules.

It would be helpful to support this conceptualization by means of a well-chosen terminology, and 'transparent' seems fine to me. So `FancyString` is transparent because we "can see" the underlying type `String`, and `IdNumber` is 'not transparent' because it is an important part of being an ID number that it does not do all the `int` things.

On top of that, we can refine: It's possible for an extension type `E` to have a representation type `R1` and a clause `implements ... R2 ...` where `R1` is a proper subtype of `R2`, e.g., because `R1` is private and `R2` defines the part of the interface of `R1` that we wish to provide to clients. We can also have a mostly-transparent extension type that adds some new members and _adapts_ others by redeclaring the given name (e.g., in order to use stricter types on some parameters of a method, or different default values, etc). Another way to be mostly transparent is to implement a type which is a proper supertype of the representation type.

My take on the refinements was that there may or may not be room for that level of detail here. If we can help developers to firmly establish the notion of an extension type as transparent or not transparent then it won't be a problem to deal with everything in between.

MaryaBelanger replied:
Ok, I think I figured out how to handle this (commit coming up!).

1. I'll include Lasse's points about the different potential `implements` scenarios with examples in the "Implements" section. Since that section comes first, this will hopefully set the precedent that we are *not* implying implementing the representation type or nothing at all are your only options. 

2. Then later on in the "Usage" section, I'll describe Erik's crucial use cases, transparent and non-transparent. I'll include information about the other `implements` situations in the "transparent" case section to not gatekeep the scenario too much. I'll generally try not to be too strict about "transparency" as a title, but more just "you can _think_ of these as transparent because..." (also based on Erik's comment here). 

Thanks for the attention to detail!

eernstg commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+A transparent extension type provides an *extended* interface to an existing type.
+You can invoke all the members of the representation type,
+plus any auxillary members you declare. The new interface is available
+to instances of the extension type.
```
It's actually about expressions and their static type. Perhaps this will do:
```suggestion
for expressions of the extension type.
```
(OK, github refuses to specify a `suggestion` here, but it's intended to replace line 267.)

MaryaBelanger commented on src/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+The inverse to transparency is restricting all members of the representation,
+providing a totally *different* interface to an existing type.
+This means simply not including it in the extension type's implements clause,
+and not declaring any of its members in its definition.
```
Thank you for clearing that up, I was also unsure :)

eernstg commented:
> make it clearer up front that this feature's main purpose is enabling JS interop

This characterization of the feature is limiting in several ways.

It is true that JS interop is a use case that has been on the agenda for a long time. However, the core nature of the extension type mechanism is that it allows the interface of an existing type to be modified (with "extend the interface" and "replace the interface" as important special cases), without incurring the cost of an actual wrapper object, and that's much broader than JS interop.

With JS interop, this means that we can perform native JS method invocations using standard Dart syntax and static type checking, and a similar description would fit other kinds of interop. With JS interop, almost all member declarations are `external`, and they are given special treatment by the compiler (such that we get the effect of performing a native JS invocation at each call site). You could say that each external member of a JS interop extension type is syntactic sugar for the low-level code that performs the native JS invocation, plus an assumption that every invocation will be inlined. In any case, JS interop uses extension types in a rather specialized manner, and they don't show many of the elements of the feature. For example, members almost never have a body, and there are few if any references to the representation object. That's basically unthinkable for an extension type that doesn't get special treatment by the compiler, and this means that JS interop may not be the perfect domain for introductory examples. It's still fine to mention JS interop as an important use case, but it is probably not important to explain the intricacies of JS-interop-extension-types here, as opposed to extension types in general.

Another use case sets out from a weakly typed object structure (say, a combination of `Map<String, dynamic>`, `List<dynamic>`, `String`, etc., created by a JSON parser). We can then use a schema describing the expected structure of that weakly typed data to create extension types (probably using code generation, but it could of course also be written manually, especially when the schema is simple). This would allow us to navigate any of these weakly typed object graphs safely, as long as they follow the schema. This use case does not rely on special treatment by the compiler, but an example might need to be somewhat large in order to make sense.

A very different case is 'newtype', where we introduce a static type distinction between values with the same representation. For example, we may wish to distinguish a value of type `Centimeter` from a value of type `Inch`, yet using a plain `int` as the representation. The `IdNumber` example is also in this category: The purpose is, among other things, to make sure that we don't consider an arbitrary `int` to be an `IdNumber`, and vice versa.

Another example is union types. [This repo](https://github.com/eernstg/extension_type_unions/tree/main) shows how extension types can be used to emulate the core properties of union types. The emulation is not nearly as convenient as a real language mechanism. Consider, say, the union of the types `int` and `String`. The emulation of `int | String` isn't equal to `String | int`, and there's a need to taken an explicit step in order to obtain an expression of type `int | String` from an expression of type `int`. However, this emulation _will_ allow a developer to define an API where a method accepts an argument of type `int` or `String`, reporting a compile-time error for all other types, and it can be done _today_ using extension types.

Another example is that it can be used to modify the signature of an existing method. The method `Future.then` accepts an argument named `onError` whose type is `Function`. This allows the `onError` argument to be _any_ function whatsoever, which is not safe (it is actually required to be something much more specific, but something which cannot be expressed as a Dart type). So the following small program fails at run time (because `print` certainly won't work):

```dart
void main() async {
  var fut = Future<int>.error("Failed!");
  await fut.then((_) => 42, onError: print); // No compile-time error; throws.
}
```

Here's a variant that uses an extension type to make the signature of `Future.then` safer:

```dart
import 'safe_future.dart';

void main() async {
  var fut = Future<int>.error("Failed!");
  // await fut.then((_) => 42, onError: print); // Compile-time error.
  int i = await fut.then((_) => 42, onError: (o) => 24); // OK!
}
```

As you can see, the unsafe invocation is now a compile-time error. The extension type looks as follows (of course, it's just an example, we'd do several additional things in order to make it realistic):

```dart
// ----- Library 'safe_future.dart'.
import 'dart:async' hide Future;
import 'dart:async' as async show Future;

extension type Future<T>._(async.Future<T> _it) implements async.Future<T> {
  Future.error(Object error, [StackTrace? stackTrace])
      : this._(async.Future.error(error, stackTrace));
  // Other constructors of `Future` are replicated similarly.
  
  Future<R> then<R>(
    FutureOr<R> onValue(T value), {
    T Function(Object)? onError,
  }) =>
      Future._(_it.then(onValue, onError: onError));

  Future<R> thenWithStack<R>(
    FutureOr<R> onValue(T value), {
    required T Function(Object, StackTrace) onError,
  }) =>
      Future._(_it.then(onValue, onError: onError));
}
```

Yet another example is that it can be used to emulate invariance of type variables. Due to dynamically checked covariance, we can easily write an example where there are no compile-time errors, and yet there's a run-time type error:

```dart
void main() {
  List<int> xs = [1];
  xs.add(1); // OK at compile time and run time.
  List<num> ys = xs; // Also OK all the way.
  ys.add(1.5); // OK at compile time, throws a type error at run time!
}
```

We could add statically checked variance (dart-lang/language#524) to the language, but we can also handle the issue today using an extension type:

```dart
// Library 'safe_list.dart'.
import 'dart:core' as core show List;
import 'dart:core' hide List;

typedef Inv<X> = X Function(X);
typedef List<X> = _List<X, Inv<X>>;

extension type const _List<X, I extends Inv<X>>(core.List<X> _it)
    implements core.List<X> {}

// Library 'my_library.dart'.
import 'safe_list.dart';

void main() {
  List<int> xs = List([1]);
  xs.add(1); // OK at compile time and run time.
  List<num> ys = xs; // Compile-time error.
  ys.add(1.5); // Irrelevant, declaration of `ys` is an error.
}
```

This illustrates yet again that extension types can be used to make adjustments to the given static typing properties of existing types, like a small type system extension engine. We may well wish to add some of those enhancements of the type system as proper language features, but surely some such cases will be application domain specific or narrow in scope, in which case they will probably never be supported as language mechanisms. In any case, it _does_ matter that we can start using an emulation of such type system enhancements already today.

Extension types is a language mechanism, and it can be used for many purposes. I'm sure we will develop more ways to use it over time, just like any other language mechanism worth its salt. It shouldn't be impossible to answer the 'why' question without cutting it down to "just for JS interop".

eernstg reviewed (commented):
Added a long comment in order to hint at answers to the 'why' question.

kevmoo commented:
Ugh! I hate to pile on last minute.

Should we discuss private constructors here?

```dart
extension type MyConsole._(JSObject thing) implements JSObject {
  static MyConsole get instance => MyConsole._(globalContext['console'] as JSObject);
  
  external void log(JSAny? value);
  external void debug(JSAny? value);
  external void info(JSAny? value);
  external void warn(JSAny? value);
}
```

kevmoo commented on src/content/language/extension-types.md:
```diff
@@ -0,0 +1,315 @@
+// 'E' has a named constructor 'n',
+// as well as its built-in unnamed constructor
+extension type E(int i) {
+  E.n(this.i, String foo);
```
Is it...weird...to have a `foo` parameter here that's not used?

This makes sense.
```dart
extension type E(int i) {
  factory E.n(String foo) => E(int.parse(foo));
}
```

MaryaBelanger replied:
Yeah, that fix is in my current WIP, yet to be committed (but so close 🥲 )

kevmoo replied:
Here you go

```dart
  E.fromString(String foo) : this.i = int.parse(foo);
```

MaryaBelanger commented:
> This characterization of the feature is limiting in several ways.

@eernstg Thanks for that illuminating write up. I've mentioned before I am hoping to write a blog post that dives into all these interesting use cases for extension types, and this is great information for that! 

I had added a sentence saying that extension types were "specifically implemented to enable static JS interop", but I've changed that now to better emphasize that they enable static js interop because of beneficial characteristics. 

Also, I unfortunately might have to merge this before you'll have time for another look, but please rereview the most recent version as I made a lot of changes based on the feedback! Thank you so much.
